File: 0200-number-of-islands/0200-number-of-islands.cpp

class Solution {
    void dfs(vector<vector<char>>& grid, int row, int column) {
        int number_of_rows = grid.size();
        int number_of_columns = grid[0].size();
        if (row < 0 || row >= number_of_rows || column < 0 || column >= number_of_columns || grid[row][column] == '0') {
            return;
        }
        grid[row][column] = '0';
        dfs(grid, row + 1, column);
        dfs(grid, row - 1, column);
        dfs(grid, row, column + 1);
        dfs(grid, row, column - 1);
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size() , m = grid[0].size();
        vector<vector<bool>> vis(n , vector<bool>(m , false));

        int ans = 0 ;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!vis[i][j] and grid[i][j] == '1') {
                    dfs(grid , i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {string[][]} grid - The 2D grid of characters.
 * @param {number} row - The current row index.
 * @param {number} column - The current column index.
 */
function dfs(grid, row, column) {
    const rows = grid.length;
    const cols = grid[0].length;

    if (row < 0 || row >= rows || column < 0 || column >= cols || grid[row][column] === '0') {
        return;
    }

    grid[row][column] = '0';

    dfs(grid, row + 1, column); // Down
    dfs(grid, row - 1, column); // Up
    dfs(grid, row, column + 1); // Right
    dfs(grid, row, column - 1); // Left
}
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
    let n = grid.length, m = grid[0].length;
    let ans = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (grid[i][j] === '1') {
                ans++;
                dfs(grid, i, j);
            }
        }
    }
    return ans;
};